### DISCUSSION TITLE
current best workaround to remap C-u to C-u zz

### DISCUSSION THREAD

**Discussion by MohamedOsman1998** [vscode-neovim/vscode-neovim#1234] [Q&A]
I understand that scrolling is still being re implemented but if there is a good workaround to center the viewport when using ctrl-u and ctrl-d please let me know, thanks

**Comment by martin-braun** [comment]
The problem is, that the cursor is not moving. You would need to move the cursor while scrolling, `zz` is wrong. Anybody?

**Reply by martin-braun** [reply]
I also tried to do what you request @MohamedOsman1998, but I can't get the `zz` mapping to work. I told it to pass `<C-d>` and `<C-u>` to Neovim, so the scrolling is closer to the original, but it fails to scroll the window now, unless the cursor gets out of sight.

Unfortunately, `10H` isn't working either, otherwise I could scroll based on `winheight(0)/4`. Mappings `<C-d>zz` should actually be enough, but mapping `zz` isn't even working.

**Reply by MohamedOsman1998** [reply]
yeah so this is a known limitation if im understanding this correctly, refer to the threads here https://github.com/vscode-neovim/vscode-neovim/pull/993, but some people apparently have some workarounds? I am just looking to get any good workaround or solution for centered scrolling at this point

**Reply by martin-braun** [reply]
@MohamedOsman1998 yup. The closest work around is to restore the Neovim behavior by passing through the key mappings using these VSCode bindings:

```json
    {
        "command": "vscode-neovim.send",
        "key": "ctrl+u",
        "when": "editorTextFocus",
        "args": "<C-u>"
    },
    {
        "command": "vscode-neovim.send",
        "key": "ctrl+d",
        "when": "editorTextFocus",
        "args": "<C-d>"
    },
```

It's way better this way, because it makes you scroll via cursor movement. Doing so will let us lose the window scrolling itself, though, unless the cursor gets out of sight.

It kinda feels the solution could be to do just both. Hey @theol0403, it's probably not that simple, or is it?

**Reply by theol0403** [reply]
Yes, the "correct" solution is to let neovim handle all the scrolling, but there are some bugs that have not been solved yet, see https://github.com/vscode-neovim/vscode-neovim/pull/885

Unfortunately, a workaround would either be:
- one of the vscode-side PRs mentioned in my PR
- creating a vscode command that calls C-u + ZZ
- dealing with the viewport height not always being in sync

**Reply by martin-braun** [reply]
@theol0403 Thanks for the insight.

> creating a vscode command that calls C-u + ZZ

I also tried to do this:

```json
    {
        "command": "vscode-neovim.send",
        "key": "ctrl+u",
        "when": "editorTextFocus",
        "args": "<C-u>zz"
    },
    {
        "command": "vscode-neovim.send",
        "key": "ctrl+d",
        "when": "editorTextFocus",
        "args": "<C-d>zz"
    },
```

That didn't help either. For some reason any viewport changes can't be sent to Neovim via `args`. I think that's the reason why you implemented a custom work around to modify the view port, which still lacks moving the cursor with it.

> dealing with the viewport height not always being in sync

That would be the best solution. Simply following my proposal to modify viewport and cursor side by side is probably a slippery slope. Are we missing any API for that? Can we poke the Neovim devs to provide one?

**Reply by theol0403** [reply]
The proper solution is https://github.com/vscode-neovim/vscode-neovim/pull/993
